iup/__init__.py

```diff
@@ -69,24 +69,58 @@ def split_train_test(
         if side == "train":
             out = (
                 pl.concat(uptake_data_list)
-                .sort("date")
-                .filter(pl.col("date") < start_date)
+                .sort("time_end")
+                .filter(pl.col("time_end") < start_date)
             )
         elif side == "test":
             out = (
                 pl.concat(uptake_data_list)
-                .sort("date")
-                .filter(pl.col("date") >= start_date)
+                .sort("time_end")
+                .filter(pl.col("time_end") >= start_date)
             )
         else:
             raise RuntimeError(f"Unrecognized side '{side}'")
 
         return out
 
+    @staticmethod
+    def date_to_season(
+        date: pl.Expr, season_start_month: int = 9, season_start_day: int = 1
+    ) -> pl.Expr:
+        """
+        Extract winter season from a date
+
+        Dates in year Y before the season start (e.g., Sep 1) are in the second part of
+        the season (i.e., in season Y-1/Y). Dates in year Y after the season start are in
+        season Y/Y+1. E.g., 2023-10-07 and 2024-04-18 are both in "2023/24"
+
+        Parameters
+        date: pl.Expr
+            dates
+        season_start_month: int
+            month of the year the season starts
+        season_start_day: int
+            day of season_start_month that the season starts
+
+        Returns
+        pl.Expr
+            seasons for each date
+        """
+
+        # for every date, figure out the season breakpoint in that year
+        season_start = pl.date(date.dt.year(), season_start_month, season_start_day)
+
+        # what is the first year in the two-year season indicator?
+        date_year = date.dt.year()
+        year1 = pl.when(date < season_start).then(date_year - 1).otherwise(date_year)
+
+        year2 = year1 + 1
+        return pl.format("{}/{}", year1, year2)
+
 
 class IncidentUptakeData(UptakeData):
     def to_cumulative(
-        self, group_cols: tuple[str,] | None, last_cumulative=None
+        self, group_cols: List[str,] | None, last_cumulative=None
     ) -> pl.DataFrame:
```

Comment:
Why do we change from tuple to List, should `group_cols` better be immutable?

Reply:
The honest answer is that I'm not (yet) a stickler for the exact right data type, as long as it works. But I do take advantage of the mutability of `group_cols` in one or two places in this code.

Reply:
These changes were part of a previous PR; I just rebased; so you can see @Fuhan-Yang that this PR is really just the 1 commit